dill/_utils.py

```diff
@@ -0,0 +1,207 @@
+
+    add = partialmethod(__proxy__, 'add')
+    discard = partialmethod(__proxy__, 'discard')
+    remove = partialmethod(__proxy__, 'remove')
+
+    def update(self, rules):
+        if isinstance(rules, ExcludeRules):
+           for filter_set in self.exclude.filter_sets:
+                getattr(self.exclude, filter_set).update(getattr(rules.exclude, filter_set))
+                getattr(self.include, filter_set).update(getattr(rules.include, filter_set))
+        else:
+            # Validate rules.
+            for rule in rules:
+                if not isinstance(rule, tuple) or len(rule) != 2:
+                    raise ValueError("invalid rule format: %r" % rule)
+            for rule_type, filter in rules:
+                if isiterable(filter):
+                    for f in filter:
+                        self.add(f, rule_type=rule_type)
+                else:
+                    self.add(filter, rule_type=rule_type)
+
+    def clear(self):
+        self.exclude.clear()
+        self.include.clear()
+
+    def filter_namespace(self, namespace, obj=None):
+        if not self.exclude and not self.include:
+            return namespace
+
+        # Protect agains dict changes during the call.
+        namespace_copy = namespace.copy() if obj is None or namespace is vars(obj) else namespace
+        objects = all_objects = [NamedObj._make(item) for item in namespace_copy.items()]
+
+        for filters in (self.exclude, self.include):
+            if filters is self.exclude and not filters:
+                # Treat the rule set as an allowlist.
+                exclude_objs = objects
+                continue
+            elif filters is self.include:
+                if not filters or not exclude_objs:
+                    break
+                objects = exclude_objs
+
+            flist = []
+            types_list = tuple(filters.types)
+            # Apply cheaper/broader filters first.
+            if types_list:
+                flist.append(lambda obj: isinstance(obj.value, types_list))
+            if filters.ids:
+                flist.append(lambda obj: id(obj.value) in filters.ids)
+            if filters.names:
+                flist.append(lambda obj: obj.name in filters.names)
+            if filters.regex:
+                flist.append(lambda obj: any(regex.fullmatch(obj.name) for regex in filters.regex))
+            flist.extend(filters.funcs)
+            for f in flist:
+                objects = filterfalse(f, objects)
+
+            if filters is self.exclude:
+                include_names = {obj.name for obj in objects}
+                exclude_objs = [obj for obj in all_objects if obj.name not in include_names]
+            else:
+                exclude_objs = list(objects)
+
+        if not exclude_objs:
+            return namespace
+        if len(exclude_objs) == len(namespace):
+            warnings.warn("filtering operation left the namespace empty!", PicklingWarning)
```

Comment:
Warning: "pickling an empty namespace" ?  Do we want a warning here?

Reply:
Rephrasing this is a must. I think yes, if some filters combination do not behave as expected and ends up excluding everything with no errors and creating an empty file, it could lead to data loss... Without this, if the user don't check the file size, they could have a bitter surprise later.